pkg/agent/openflow/client.go

```diff
@@ -233,6 +236,19 @@ func (c *client) UninstallPodFlows(containerID string) error {
 	return c.deleteFlows(c.podFlowCache, containerID)
 }
 
+func (c *client) GetPodFlows(containerID string) []binding.Flow {
+	fCacheI, ok := c.podFlowCache.Load(containerID)
+	if !ok {
+		return nil
+	}
+	fCache := fCacheI.(flowCache)
+	flows := make([]binding.Flow, 0, len(fCache))
+	for _, flow := range fCache {
+		flows = append(flows, flow)
+	}
+	return flows
```

Comment:
I don't think there is an actual issue right now, but we have to be careful when we return `binding.Flow` objects (since it is an interface). There could be an issue caused by concurrent access. One possibility is if `GetPodFlows` is called (and the return value is consumed) while a call to `ReplayFlows` is happening. That's because `ReplayFlows` is modifying the flow objects stored in the cache (by calling `flow.Reset()`). IIRC `Reset` only modifies one field so things should be good as long as the caller of `GetPodFlows` does not access that field. But this is hard to guarantee moving forward. Maybe `GetPodFlows` should instead return a copy of the flows, or take care of the serialization itself, and acquire a read lock on `c.replayMutex`.

I believe that this issue only happens with `ReplayFlows`, not with the other function which access the flow cache (`addFlows` and `deleteFlows`).

Reply:
I did not know this before. Will check, but prefer to acquire a read lock and return a list of flow match strings instead.
Or do you think in ReplyFlows we can change to make a copy the Flows to modify?

Reply:
Usually my preference is to make the copy in the "get" function. This prevents the client from actually modifying your internal data structure.

Reply:
Could you remind me why we call Flow.Reset()? Seems Reset() only does: f.Flow.Table = f.table.Table.

Reply:
IIRC, when there is an OVS reconnect, the ofnet library creates new OFTable objects. That means that our flow cache essentially becomes invalid (and we cannot "add" the flows again) until references to the old tables (`f.Flow.Table`) in each cached flow have been replaced with references to the new tables.

Reply:
Note that the client cannot access `f.Flow` directly anyway, and `MatchString` does not access `f.Flow.Table`, only `f.Flow.Match`, which is why I said there wasn't an *actual* race-condition. But I do feel making a copy in `GetPodFlows` is the right way to go.

Reply:
Many our codes assume the set/list can be changed, but the object itself is immutable. It is a common pattern in K8s controller implementation to achieve lockless and better performance. In this Flow case, I feel we can still return the cached Flow, but in some way mark the Flow.Flow and Flow.table as internal and mutable. But I know you might think this is not clear enough and might not be easy to track. So, I am open to hear your thoughts. @tnqn @wenyingd

Reply:
How do you feel about using interface composition and returning a different interface here, so that we limit the operations the client can do with the returned flows?
```
type FlowReader interface {
        MatchString() string
}
type Flow interface {
	OFEntry
	FlowReader
	// CopyToBuilder returns a new FlowBuilder that copies the matches of the Flow, but does not copy the actions.
	CopyToBuilder() FlowBuilder
}
```

But I don't feel too strongly about this: there is no race condition with the current usage, and this is still an "internal" interface that we consume in Antrea, so we control how it is used. But a comment in `GetPodFlows` would be nice if you don't change anything.

Reply:
I changed the func to return matchStrings, with the relayMutex read lock. Also added comments about the table change at reconnection/replay. Please check.